change stacksize_analysis to worklist algorithm for better result

zyq8709 opened the pull request:
The current algorithm for stacksize_analysis is round-robin algorithm with 100 as iteration upper limit, which is not an optimal choice. This PR implements the worklist version for better complexity and precision.


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng @soumith @yanboliang @anijain2305 @desertfire

zyq8709 committed:
change stacksize_analysis to worklist algorithm for better result

jansel reviewed (changes requested):
Changes seem fine, but it looks like tests are failing in CI.

zyq8709 commented:
Fix some errors and re-commit. Still 2 fails. Seems weird and irrelevant.

jansel commented on torch/_dynamo/bytecode_analysis.py:
```diff
@@ -132,46 +133,52 @@ class FixedPointBox:
-                stack_sizes[inst.target].offset_of(
-                    stack_size, stack_effect(inst.opcode, inst.arg, jump=True)
-                )
+    worklist = list()
```
Suggest using `collections.deque()`, the `pop(0)` below is `O(n)` with a list, `O(1)` with a deque.

zyq8709 replied:
Thanks for your advice! I will change it accordingly.

jansel commented on torch/_dynamo/bytecode_analysis.py:
```diff
@@ -181,6 +188,6 @@ def stacksize_analysis(instructions):
     high = max([x.high for x in stack_sizes.values()])
 
-    assert fixed_point.value, "failed to reach fixed point"
+    fixed_point.value = True
```
Unused?

zyq8709 replied:
> Unused?

Yes, just to confirm fixed point is reached

jansel commented on torch/_dynamo/bytecode_analysis.py:
```diff
@@ -132,46 +133,52 @@ class FixedPointBox:
+            stack_sizes[instructions[index + 1]].offset_of(
+                stack_size, stack_effect(inst.opcode, inst.arg, jump=False), fixed_point
+            )
+            if not fixed_point.value:
```
With this change you can actually just delete `FixedPointBox()` and instead make `offset_of()` return it.

zyq8709 replied:
Got it. I just intended to reuse the original codes. Then I will rewrite this.

jansel reviewed (changes requested):
Those timeouts in CI are a bit worrying.  I wonder if this could be going into an infinite loop to cause them.

zyq8709 commented:
> Those timeouts in CI are a bit worrying. I wonder if this could be going into an infinite loop to cause them.
I was also concerned about these failures and added some log outputs. Based on the log, each loop had exited normally. I don't know how to debug CI remotely. Maybe log is not enough.

zyq8709 committed:
remove FixedPointBox and use deque instead of list

jansel commented:
> > Those timeouts in CI are a bit worrying. I wonder if this could be going into an infinite loop to cause them.
> 
> I was also concerned about these failures and added some log outputs. Based on the log, each loop had exited normally. I don't know how to debug CI remotely. Maybe log is not enough.
> 
> Regarding general interval analysis for variables' value, it is necessary to use widening for fixed point acceleration. However, for the interval of stack size it is a little special, which should be actually finite and not a very big range, although the lattice is almost infinite. Theoretically, the transfer functions are monotone in this model and fixed point can be reached. It is wired that it takes too long to stabilize under some extreme circumstances perhaps due to the large lattice.

Perhaps add some asserts if the algorithm has run for some huge number of iterations and see if CI trips them.

zyq8709 commented:
> > > Those timeouts in CI are a bit worrying. I wonder if this could be going into an infinite loop to cause them.
> > 
> > 
> > I was also concerned about these failures and added some log outputs. Based on the log, each loop had exited normally. I don't know how to debug CI remotely. Maybe log is not enough.
> > Regarding general interval analysis for variables' value, it is necessary to use widening for fixed point acceleration. However, for the interval of stack size it is a little special, which should be actually finite and not a very big range, although the lattice is almost infinite. Theoretically, the transfer functions are monotone in this model and fixed point can be reached. It is wired that it takes too long to stabilize under some extreme circumstances perhaps due to the large lattice.
> 
> Perhaps add some asserts if the algorithm has run for some huge number of iterations and see if CI trips them.

Seems that it is largely caused by unstable running of CI. All passed this time...

jansel commented on torch/_dynamo/bytecode_analysis.py:
```diff
@@ -123,55 +125,57 @@ def walk(state, start):
+    count = 0
+    while len(worklist) != 0:
+        count += 1
+        assert count < 10000, "Number of iterations has exceeded the limit"
```
lgtm, let's remove this line before merging though.

zyq8709 replied:
Have dropped the 'CI assertion' commit

jansel reviewed (changes requested):
It looks like the CI failures are back.

I suspect it passed with the assertion due to `dynamo.config.suppress_errors`.  Without the assert it infinite loops.  

To reproduce you are going to want to set `PYTORCH_TEST_WITH_DYNAMO=1` and run the file that is hanging.

zyq8709 commented:
Seems failure cases changed.  But time out cases [linux-bionic-py3.8-clang9 / test (dynamo, 2, 2, linux.2xlarge)](https://github.com/pytorch/pytorch/actions/runs/4433980518/jobs/7779681273#logs) and [linux-bionic-py3.8-clang9 / test (dynamo, 2, 2, linux.2xlarge)](https://github.com/pytorch/pytorch/actions/runs/4433980518/jobs/7779681273#logs) also passed this time.

jansel commented:
When `suppress_errors=True` (the default for the PYTORCH_TEST_WITH_DYNAMO=1 tests) then `AssertionError` results in falling back to eager mode and the test passing.

If we hit the assertion you could print out the problematic IR then call `sys.exit(1)` to get a better sense of the problematic bytecode.

ar0ck commented on torch/_dynamo/bytecode_analysis.py:
```diff
@@ -164,55 +165,55 @@ def walk(state, start):
+            changed = stack_sizes[inst.target].offset_of(
+                stack_size, stack_effect(inst.opcode, inst.arg, jump=True)
+            )
+            if changed:
```
[Here](https://github.com/pytorch/pytorch/issues/97811#issuecomment-1591888590) is a repro which causes your branch to loop infinitely (with `python3.8`).

This is an existing bug (#97811) which I was trying to fix (#103621), but it seems easier to address here.

The issue is that we are not correctly modelling the `CALL_FINALLY` opcode - see the `cpython3.8` implementation [here](https://github.com/python/cpython/blob/3.8/Python/compile.c#L5499).

I think this patch should solve the issue:

```suggestion
            if changed and (
                sys.version_info >= (3, 9) or inst.opcode != dis.opmap["CALL_FINALLY"]
            ):
```

zyq8709 replied:
Sorry for the late reply! Thanks a lot for your suggestion! I will check this later. Maybe this PR can be merged after the fix.

ar0ck replied:
Sounds good - in the meantime, #103621 did go in, so rebasing might have fixed it. That being said, I think doing it this way lines up better with the CPython implementation.

zyq8709 committed:
Update torch/_dynamo/bytecode_analysis.py

Co-authored-by: Andy Rock <7538433+ar0ck@users.noreply.github.com>
